Add support for Open Enclave 0.10.0

[Open Enclave 0.10.0](https://github.com/openenclave/openenclave/releases/tag/v0.10.0) [deprecates](https://github.com/openenclave/openenclave/issues/2595) the `oehostapp` CMake target in favor of `oehost`. However, simply changing our CMake targets from `oehostapp` to `oehost` causes compilation errors:

```
CMakeFiles/xgboost_enclave.dir/rabit/src/ssl_socket.cc.o: In function `rabit::utils::SSLTcpSocket::ConfigureClientSSL()':
ssl_socket.cc:(.text._ZN5rabit5utils12SSLTcpSocket18ConfigureClientSSLEv+0xbd): undefined reference to `mbedtls_debug_set_threshold'
CMakeFiles/xgboost_enclave.dir/rabit/src/ssl_socket.cc.o: In function `rabit::utils::SSLTcpSocket::ConfigureServerSSL()':
ssl_socket.cc:(.text._ZN5rabit5utils12SSLTcpSocket18ConfigureServerSSLEv+0x147): undefined reference to `mbedtls_debug_set_threshold'
/opt/openenclave/lib/openenclave/enclave/liboecore.a(cpuid.c.o): In function `oe_initialize_cpuid':
/source/enclave/core/sgx/cpuid.c:29: undefined reference to `oe_sgx_get_cpuid_table_ocall'
collect2: error: ld returned 1 exit status
enclave/CMakeFiles/xgboost_enclave.dir/build.make:960: recipe for target 'enclave/xgboost_enclave' failed
make[2]: *** [enclave/xgboost_enclave] Error 1
CMakeFiles/Makefile2:349: recipe for target 'enclave/CMakeFiles/xgboost_enclave.dir/all' failed
make[1]: *** [enclave/CMakeFiles/xgboost_enclave.dir/all] Error 2
Makefile:129: recipe for target 'all' failed
make: *** [all] Error 2
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -21,7 +21,7 @@ before_install:
   - wget -qO - https://packages.microsoft.com/keys/microsoft.asc | sudo apt-key add -
   - sudo apt-get -qq update
   # Install dependencies
-  - sudo apt -y install clang-7 libssl-dev gdb libsgx-enclave-common libsgx-enclave-common-dev libprotobuf10 libsgx-dcap-ql libsgx-dcap-ql-dev az-dcap-client open-enclave=0.9.0 libmbedtls-dev
+  - sudo apt -y install clang-7 libssl-dev gdb libsgx-enclave-common libsgx-enclave-common-dev libprotobuf10 libsgx-dcap-ql libsgx-dcap-ql-dev az-dcap-client open-enclave=0.12.0 libmbedtls-dev
   - wget https://github.com/Kitware/CMake/releases/download/v3.15.6/cmake-3.15.6-Linux-x86_64.sh
   - sudo bash cmake-3.15.6-Linux-x86_64.sh --skip-license --prefix=/usr/local
   - export PATH=/usr/local:/usr/local/bin:$PATH 

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -112,6 +112,14 @@ endif(LVI_MITIGATION)
 
 set_default_configuration_release()
 
+set(OE_MIN_VERSION 0.12.0)
+if (LVI_MITIGATION)
+    # Configure the cmake to use customized compilation toolchain.
+    # This package has to be added before `project()`.
+    find_package(OpenEnclave-LVI-Mitigation ${OE_MIN_VERSION} CONFIG REQUIRED)
+endif()
+find_package(OpenEnclave ${OE_MIN_VERSION} CONFIG REQUIRED)
+
 add_subdirectory(${PROJECT_SOURCE_DIR}/host)
 add_subdirectory(${PROJECT_SOURCE_DIR}/enclave)
 

--- a/README.md
+++ b/README.md
@@ -39,10 +39,10 @@ In Secure XGBoost, we design and implement data-oblivious algorithms for model t
 Unfortunately, the extra security comes at the cost of performance. If such attacks fall outside the users’ threat model, they can disable this extra protection.
 
 ## Installation
-1. Install the Open Enclave SDK (0.9.0) and the Intel SGX DCAP driver by following [these instructions](https://github.com/openenclave/openenclave/blob/master/docs/GettingStartedDocs/install_oe_sdk-Ubuntu_18.04.md). In Step 3 of the instructions, install Open Enclave version 0.9.0 by specifying the version:
+1. Install the Open Enclave SDK (0.12.0) and the Intel SGX DCAP driver by following [these instructions](https://github.com/openenclave/openenclave/blob/master/docs/GettingStartedDocs/install_oe_sdk-Ubuntu_18.04.md). In Step 3 of the instructions, install Open Enclave version 0.12.0 by specifying the version:
 
     ```sh
-    sudo apt -y install clang-7 libssl-dev gdb libsgx-enclave-common libsgx-enclave-common-dev libprotobuf10 libsgx-dcap-ql libsgx-dcap-ql-dev az-dcap-client open-enclave=0.9.0
+    sudo apt -y install clang-7 libssl-dev gdb libsgx-enclave-common libsgx-enclave-common-dev libprotobuf10 libsgx-dcap-ql libsgx-dcap-ql-dev az-dcap-client open-enclave=0.12.0
     ```
 
 2. Configure the required environment variables.

--- a/doc/build.rst
+++ b/doc/build.rst
@@ -28,14 +28,14 @@ to open an issue on `GitHub <https://github.com/mc2-project/secure-xgboost/issue
 Installing the Open Enclave SDK
 *******************************
 
-1. Install the Open Enclave SDK (v0.9.0) and the Intel SGX DCAP driver.  
+1. Install the Open Enclave SDK (v0.12.0) and the Intel SGX DCAP driver.  
    If building on an SGX-enabled machine, follow the instructions `here <https://github.com/openenclave/openenclave/blob/master/docs/GettingStartedDocs/install_oe_sdk-Ubuntu_18.04.md>`_. 
    
-   **Note**: In step 3 of the instructions, make sure that you install Open Enclave version 0.9.0 by specifying the version
+   **Note**: In step 3 of the instructions, make sure that you install Open Enclave version 0.12.0 by specifying the version
 
    .. code-block:: bash
 
-      sudo apt -y install clang-7 libssl-dev gdb libprotobuf10 libsgx-enclave-common libsgx-enclave-common-dev libsgx-dcap-ql libsgx-dcap-ql-dev az-dcap-client open-enclave=0.9.0
+      sudo apt -y install clang-7 libssl-dev gdb libprotobuf10 libsgx-enclave-common libsgx-enclave-common-dev libsgx-dcap-ql libsgx-dcap-ql-dev az-dcap-client open-enclave=0.12.0
 
    .. note:: You may also build the SDK in "simulation mode" on a machine without SGX support (e.g., for local development and testing). To build in simulation mode, follow the instructions `here <https://github.com/openenclave/openenclave/blob/master/docs/GettingStartedDocs/install_oe_sdk-Simulation.md>`_ instead. Notably, these instructions require that you skip the driver installation step.
 

--- a/enclave/CMakeLists.txt
+++ b/enclave/CMakeLists.txt
@@ -32,14 +32,6 @@ int main() {
 }
 " XGBOOST_BUILTIN_PREFETCH_PRESENT)
 
-set(OE_MIN_VERSION 0.8.2)
-if (LVI_MITIGATION)
-  # Configure the cmake to use customized compilation toolchain.
-  # This package has to be added before `project()`.
-  find_package(OpenEnclave-LVI-Mitigation ${OE_MIN_VERSION} CONFIG REQUIRED)
-endif()
-find_package(OpenEnclave ${OE_MIN_VERSION} CONFIG REQUIRED)
-
 # Check endianness
 include(TestBigEndian)
 test_big_endian(BIG_ENDIAN)
@@ -56,14 +48,18 @@ if (CONSENSUS)
     list(REMOVE_ITEM CPU_SOURCES ${PROJECT_SOURCE_DIR}/src/c_api/c_api.cc)
     add_custom_command(OUTPUT xgboost_mc_t.h xgboost_mc_t.c xgboost_mc_args.h 
         DEPENDS ${PROJECT_SOURCE_DIR}/xgboost_mc.edl
-        COMMAND openenclave::oeedger8r --trusted ${PROJECT_SOURCE_DIR}/xgboost_mc.edl)
+        COMMAND openenclave::oeedger8r --trusted ${PROJECT_SOURCE_DIR}/xgboost_mc.edl
+        --search-path ${OE_INCLUDEDIR} --search-path
+        ${OE_INCLUDEDIR}/openenclave/edl/sgx)
 
     add_executable(xgboost_enclave ${PROJECT_SOURCE_DIR}/ecalls_mc.cpp ${CPU_SOURCES} ${RABIT_SOURCES} ${CMAKE_CURRENT_BINARY_DIR}/xgboost_mc_t.c)
 else()
     list(REMOVE_ITEM CPU_SOURCES ${PROJECT_SOURCE_DIR}/src/c_api/c_api_mc.cc)
     add_custom_command(OUTPUT xgboost_t.h xgboost_t.c xgboost_args.h 
         DEPENDS ${PROJECT_SOURCE_DIR}/xgboost.edl
-        COMMAND openenclave::oeedger8r --trusted ${PROJECT_SOURCE_DIR}/xgboost.edl)
+        COMMAND openenclave::oeedger8r --trusted ${PROJECT_SOURCE_DIR}/xgboost.edl
+        --search-path ${OE_INCLUDEDIR} --search-path
+        ${OE_INCLUDEDIR}/openenclave/edl/sgx)
 
     add_executable(xgboost_enclave ${PROJECT_SOURCE_DIR}/ecalls.cpp ${CPU_SOURCES} ${RABIT_SOURCES} ${CMAKE_CURRENT_BINARY_DIR}/xgboost_t.c)
 endif()

--- a/enclave/rabit/src/ssl_socket.cc
+++ b/enclave/rabit/src/ssl_socket.cc
@@ -51,7 +51,6 @@ bool SSLTcpSocket::ConfigureClientSSL() {
   mbedtls_ssl_conf_rng(&conf, mbedtls_ctr_drbg_random, &ctr_drbg);
 
   mbedtls_ssl_conf_dbg( &conf, my_debug, stdout );
-  mbedtls_debug_set_threshold(DEBUG_LEVEL);
 
   // set up SSL context
   if ((ret = mbedtls_ssl_setup(&ssl, &conf)) != 0) {
@@ -115,7 +114,6 @@ bool SSLTcpSocket::ConfigureServerSSL() {
   mbedtls_ssl_conf_rng( &conf, mbedtls_ctr_drbg_random, &ctr_drbg );
 
   mbedtls_ssl_conf_dbg( &conf, my_debug, stdout );
-  mbedtls_debug_set_threshold(DEBUG_LEVEL);
 
   if ((ret = mbedtls_ssl_setup(&ssl, &conf)) != 0) {
     print_err(ret);

--- a/enclave/rabit/src/ssl_socket.h
+++ b/enclave/rabit/src/ssl_socket.h
@@ -11,7 +11,6 @@
 #include "mbedtls/ctr_drbg.h"
 #include "mbedtls/platform.h"
 #include "mbedtls/error.h"
-#include "mbedtls/debug.h"
 
 #include <memory>
 #include <string>
@@ -28,8 +27,6 @@ static void print_err(int error_code) {
   mbedtls_strerror(error_code, err_buf, LEN);
   LOG(FATAL) << err_buf;
 }
-
-#define DEBUG_LEVEL 0
 
 static void my_debug( void *ctx, int level,
     const char *file, int line,

--- a/enclave/xgboost.edl
+++ b/enclave/xgboost.edl
@@ -1,6 +1,10 @@
 enclave {
+    from "openenclave/edl/syscall.edl" import *;
+    from "platform.edl" import *;
+
     include "xgboost/c_api.h"
     include "rabit/c_api.h"
+
     trusted {
         public void enclave_init(
                 [in, count=num_users] char **usernames,

--- a/enclave/xgboost_mc.edl
+++ b/enclave/xgboost_mc.edl
@@ -1,6 +1,10 @@
 enclave {
+    from "openenclave/edl/syscall.edl" import *;
+    from "platform.edl" import *;
+
     include "xgboost/c_api_mc.h"
     include "rabit/c_api.h"
+
     trusted {
         public void enclave_init(
                 [in, count=num_users] char **usernames,

--- a/host/CMakeLists.txt
+++ b/host/CMakeLists.txt
@@ -1,7 +1,5 @@
 cmake_minimum_required(VERSION 3.11)
 project(xgboost LANGUAGES CXX C VERSION 0.90)
-# openenclave
-find_package(OpenEnclave CONFIG REQUIRED)
 
 ## Deprecation warning
 if (USE_AVX)
@@ -19,11 +17,15 @@ endif (USE_SANITIZER)
 if (CONSENSUS)
     add_custom_command(OUTPUT xgboost_mc_u.h xgboost_mc_u.c xgboost_mc_args.h 
         DEPENDS ${CMAKE_SOURCE_DIR}/enclave/xgboost_mc.edl
-        COMMAND openenclave::oeedger8r --untrusted ${CMAKE_SOURCE_DIR}/enclave/xgboost_mc.edl)
+        COMMAND openenclave::oeedger8r --untrusted ${CMAKE_SOURCE_DIR}/enclave/xgboost_mc.edl
+        --search-path ${OE_INCLUDEDIR} --search-path
+        ${OE_INCLUDEDIR}/openenclave/edl/sgx)
 else()
     add_custom_command(OUTPUT xgboost_u.h xgboost_u.c xgboost_args.h 
         DEPENDS ${CMAKE_SOURCE_DIR}/enclave/xgboost.edl
-        COMMAND openenclave::oeedger8r --untrusted ${CMAKE_SOURCE_DIR}/enclave/xgboost.edl)
+        COMMAND openenclave::oeedger8r --untrusted ${CMAKE_SOURCE_DIR}/enclave/xgboost.edl
+        --search-path ${OE_INCLUDEDIR} --search-path
+        ${OE_INCLUDEDIR}/openenclave/edl/sgx)
 endif()
 
 # dmlc-core
@@ -58,7 +60,9 @@ endif ()
 target_include_directories(rabit PRIVATE
   $<BUILD_INTERFACE:${CMAKE_SOURCE_DIR}/include>
   $<BUILD_INTERFACE:${CMAKE_SOURCE_DIR}/include/rabit>)
-target_link_libraries(rabit PRIVATE ${LINKED_LIBRARIES_PRIVATE} openenclave::oehostapp mbedcrypto mbedtls)
+target_include_directories(rabit
+    PRIVATE ${CMAKE_CURRENT_BINARY_DIR})
+target_link_libraries(rabit PRIVATE ${LINKED_LIBRARIES_PRIVATE} openenclave::oehost mbedcrypto mbedtls)
 set_target_properties(rabit
   PROPERTIES
   CXX_STANDARD 11
@@ -79,6 +83,6 @@ target_include_directories(xgboost
   $<BUILD_INTERFACE:${CMAKE_SOURCE_DIR}/include>)
 target_include_directories(xgboost
   PRIVATE ${CMAKE_CURRENT_BINARY_DIR})
-target_link_libraries(xgboost PRIVATE ${LINKED_LIBRARIES_PRIVATE} openenclave::oehostapp mbedcrypto mbedtls)
+target_link_libraries(xgboost PRIVATE ${LINKED_LIBRARIES_PRIVATE} openenclave::oehost mbedcrypto mbedtls)
 
 set_output_directory(xgboost ${CMAKE_SOURCE_DIR}/build)

--- a/host/src/CMakeLists.txt
+++ b/host/src/CMakeLists.txt
@@ -38,11 +38,15 @@ endif (PLUGIN_DENSE_PARSER)
 if (CONSENSUS)
     add_custom_command(OUTPUT xgboost_mc_u.h xgboost_mc_u.c xgboost_mc_args.h 
         DEPENDS ${CMAKE_SOURCE_DIR}/enclave/xgboost_mc.edl
-        COMMAND openenclave::oeedger8r --untrusted ${CMAKE_SOURCE_DIR}/enclave/xgboost_mc.edl)
+        COMMAND openenclave::oeedger8r --untrusted ${CMAKE_SOURCE_DIR}/enclave/xgboost_mc.edl
+        --search-path ${OE_INCLUDEDIR} --search-path
+        ${OE_INCLUDEDIR}/openenclave/edl/sgx)
 else()
     add_custom_command(OUTPUT xgboost_u.h xgboost_u.c xgboost_args.h 
         DEPENDS ${CMAKE_SOURCE_DIR}/enclave/xgboost.edl
-        COMMAND openenclave::oeedger8r --untrusted ${CMAKE_SOURCE_DIR}/enclave/xgboost.edl)
+        COMMAND openenclave::oeedger8r --untrusted ${CMAKE_SOURCE_DIR}/enclave/xgboost.edl
+        --search-path ${OE_INCLUDEDIR} --search-path
+        ${OE_INCLUDEDIR}/openenclave/edl/sgx)
 endif()
 
 
@@ -56,7 +60,7 @@ else ()
 endif ()
   # add_library(objxgboost STATIC ${CPU_SOURCES} ${PLUGINS_SOURCES})
   # add_library(objxgboost SHARED ${CPU_SOURCES} ${PLUGINS_SOURCES})
-  target_link_libraries(objxgboost openenclave::oehostapp ${LINKED_LIBRARIES_PRIVATE})
+  target_link_libraries(objxgboost openenclave::oehost ${LINKED_LIBRARIES_PRIVATE})
 
 target_include_directories(objxgboost
   PRIVATE
@@ -120,4 +124,4 @@ endif (USE_OPENMP)
 msvc_use_static_runtime()
 
 #-- End object library
-# target_link_libraries(objxgboost openenclave::oehostapp)
+# target_link_libraries(objxgboost openenclave::oehost)

